File: week-06/ek_1926_drawing.py
import sys
from collections import deque
read = sys.stdin.readline

dx = (1,0,-1,0)
dy = (0,1,0,-1)

def is_valid(x: int, y: int) -> bool:
    return False if x < 0 or x >= n or y < 0 or y >= m else True

def bfs(x: int, y: int) -> int:
    dq = deque([(x,y)])
    visited[x][y] = True # 현재 칸 방문 표시
    area = 1   # 현재 칸 포함
    
    while dq:
        sx, sy = dq.popleft()

        for k in range(4):
            nx, ny = sx + dx[k], sy + dy[k]

            if not is_valid(nx,ny) or visited[nx][ny] or not arr[nx][ny]:
                continue

            area += 1
            visited[nx][ny] = True
            dq.append((nx,ny))

    return area


n, m = map(int, read().split())
arr = [list(map(int, read().split())) for _ in range(n)]
visited = [ [False] * m for r_ in range(n)]

count = 0
max_area = 0

for row in range(n):
    for col in range(m):
        if not visited[row][col] and arr[row][col]:
            max_area = max(max_area, bfs(row, col))
            count += 1

print(count)
print(max_area)

File: week-06/ek_1926_drawing.cpp
#include <iostream>
#include <queue>
#include <cmath>
#define MAX_NUM 502
using namespace std;

int n,m;
int dx[4] = {1,0,-1,0};
int dy[4] = {0,1,0,-1};
vector<vector<bool>> visited;
vector<vector<int>> arr;

bool IsValid(int x, int y) {
    if (x < 0 || x >= n || y < 0 || y >= m)
        return false;
    return true;
}

int BFS(int sx, int sy) {
    queue<pair<int, int>> q;
    // 현재 칸 처리
    q.push({sx, sy});
    visited[sx][sy] = true;
    int area = 1;

    while (!q.empty()) {
        int x = q.front().first;
        int y = q.front().second;
        q.pop();

        for (int k = 0; k < 4; k++) {
            int nx = x + dx[k];
            int ny = y + dy[k];
            
            if (IsValid(nx, ny) && arr[nx][ny] && !visited[nx][ny]) {
                q.push({nx,ny});
                visited[nx][ny] = true;
                area++;
            }
        }
    }
    
    return area;
}

int main(void) {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    cin >> n >> m;
    
    arr.assign(n, vector<int>(m));
    visited.assign(n, vector<bool>(m, false));

    for (int row = 0; row < n; row++) {
        for (int col = 0; col < m; col++) 
            cin >> arr[row][col];
    }

    int count = 0;
    int maxArea = 0;
    
    for (int row = 0; row < n; row++) {
        for (int col = 0; col < m; col++) {
            if(arr[row][col] && !visited[row][col]) {
                maxArea = max(maxArea, BFS(row, col));
                count++;
            }
        }
    }

    cout << count << "\n" << maxArea;

    return 0;
}